### ISSUE TITLE
App Passwords (when using 2StepAuth) can lock a user if there is a policy for that

### ISSUE THREAD

**Issue by r3pek** [nextcloud/server#2999]
### Steps to reproduce
1. Install and configure 2StepAuth plugin
2. Install and configure auth_ldap plugin
3. setup some policies on the ldap server (FreeIPA in this case), to lock accounts after X failed attempts
4. Create an App Password
5. Try to login with that app password

### Expected behaviour
Apps should login fine

### Actual behaviour
Apps don't login and the "main" accounts get's locked for invalid passwords. 
I tested this forcing a sync on and Android phone runing the latest DAVDroid.
problem didn't occur *before* changing the auth to LDAP.

### Server configuration
**Operating system**: Arch Linux

**Web server:** Nginx

**Database:** MariaDB

**PHP version:** 7.0.14

**Nextcloud version:** (see Nextcloud admin page) 10.0.2

**Updated from an older Nextcloud/ownCloud or fresh install:** comes from OwnCloud

**Where did you install Nextcloud from:** Downloads from main site.

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
  - activity: 2.3.2
  - admin_audit: 1.0.0
  - calendar: 1.4.1
  - comments: 1.0.0
  - contacts: 1.5.1
  - dav: 1.0.1
  - federatedfilesharing: 1.0.1
  - federation: 1.0.1
  - files: 1.5.2
  - files_accesscontrol: 1.1.2
  - files_automatedtagging: 1.1.1
  - files_pdfviewer: 0.8.1
  - files_sharing: 1.0.0
  - files_texteditor: 2.1
  - files_trashbin: 1.0.0
  - files_versions: 1.3.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - gallery: 15.0.1
  - news: 9.0.4
  - notes: 2.0.2
  - notifications: 0.3.0
  - password_policy: 1.0.0
  - passwords: 19
  - provisioning_api: 1.0.0
  - qownnotesapi: 16.09.0
  - serverinfo: 1.1.1
  - shorten: 0.0.15
  - survey_client: 0.1.5
  - systemtags: 1.0.2
  - templateeditor: 0.1
  - theming: 1.0.1
  - twofactor_totp: 0.4.1
  - updatenotification: 1.0.1
  - user_ldap: 1.0.1
  - workflowengine: 1.0.1
```
</details>

**The content of config/config.php:**
<details>
<summary>Config report</summary>

```
    "system": {
        "instanceid": "ocg4w2lbke5t",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.r3pek.org"
        ],
        "datadirectory": "\/storage\/nextcloud\/data",
        "overwrite.cli.url": "https:\/\/cloud.r3pek.org",
        "dbtype": "mysql",
        "version": "9.1.2.2",
        "dbname": "owncloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "Europe\/Lisbon",
        "installed": true,
        "appstore.experimental.enabled": true,
        "maintenance": false,
        "loglevel": 2,
        "htaccess.RewriteBase": "\/",
        "theme": "",
        "updater.server.url": "https:\/\/updates.nextcloud.com\/updater_server\/",
        "updater.release.channel": "stable",
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory"
    }
}
```

</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** LDAP

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
+-------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------+
| Configuration                 |                                                                                                                                                   |
+-------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------+
| hasMemberOfFilterSupport      | 1                                                                                                                                                 |
| hasPagedResultSupport         |                                                                                                                                                   |
| homeFolderNamingRule          | attr:uid                                                                                                                                          |
| lastJpegPhotoLookup           | 0                                                                                                                                                 |
| ldapAgentName                 | uid=nextcloud,cn=sysaccounts,cn=etc,dc=r3pek,dc=org                                                                                               |
| ldapAgentPassword             | ***                                                                                                                                               |
| ldapAttributesForGroupSearch  |                                                                                                                                                   |
| ldapAttributesForUserSearch   |                                                                                                                                                   |
| ldapBackupHost                |                                                                                                                                                   |
| ldapBackupPort                |                                                                                                                                                   |
| ldapBase                      | cn=accounts,dc=r3pek,dc=org                                                                                                                       |
| ldapBaseGroups                | cn=groups,cn=accounts,dc=r3pek,dc=org                                                                                                             |
| ldapBaseUsers                 | cn=users,cn=accounts,dc=r3pek,dc=org                                                                                                              |
| ldapCacheTTL                  | 600                                                                                                                                               |
| ldapConfigurationActive       | 1                                                                                                                                                 |
| ldapDynamicGroupMemberURL     |                                                                                                                                                   |
| ldapEmailAttribute            | mail                                                                                                                                              |
| ldapExperiencedAdmin          | 1                                                                                                                                                 |
| ldapExpertUUIDGroupAttr       | ipaUniqueID                                                                                                                                       |
| ldapExpertUUIDUserAttr        | ipaUniqueID                                                                                                                                       |
| ldapExpertUsernameAttr        |                                                                                                                                                   |
| ldapGroupDisplayName          | cn                                                                                                                                                |
| ldapGroupFilter               | (&(|(objectclass=groupofnames))(|(memberOf=cn=nextcloud,cn=groups,cn=accounts,dc=r3pek,dc=org)))                                                  |
| ldapGroupFilterGroups         | nextcloud                                                                                                                                         |
| ldapGroupFilterMode           | 1                                                                                                                                                 |
| ldapGroupFilterObjectclass    | groupofnames                                                                                                                                      |
| ldapGroupMemberAssocAttr      | member                                                                                                                                            |
| ldapHost                      | srv04.r3pek.org                                                                                                                                   |
| ldapIgnoreNamingRules         |                                                                                                                                                   |
| ldapLoginFilter               | (&(&(|(objectclass=person))(|(memberof=cn=nextcloud,cn=groups,cn=accounts,dc=r3pek,dc=org)))(|(uid=%uid)(|(mailPrimaryAddress=%uid)(mail=%uid)))) |
| ldapLoginFilterAttributes     |                                                                                                                                                   |
| ldapLoginFilterEmail          | 1                                                                                                                                                 |
| ldapLoginFilterMode           | 0                                                                                                                                                 |
| ldapLoginFilterUsername       | 1                                                                                                                                                 |
| ldapNestedGroups              | 0                                                                                                                                                 |
| ldapOverrideMainServer        |                                                                                                                                                   |
| ldapPagingSize                | 500                                                                                                                                               |
| ldapPort                      | 389                                                                                                                                               |
| ldapQuotaAttribute            |                                                                                                                                                   |
| ldapQuotaDefault              |                                                                                                                                                   |
| ldapTLS                       | 0                                                                                                                                                 |
| ldapUserDisplayName           | displayname                                                                                                                                       |
| ldapUserDisplayName2          |                                                                                                                                                   |
| ldapUserFilter                | (&(|(objectclass=person))(|(memberof=cn=nextcloud,cn=groups,cn=accounts,dc=r3pek,dc=org)))                                                        |
| ldapUserFilterGroups          | nextcloud                                                                                                                                         |
| ldapUserFilterMode            | 0                                                                                                                                                 |
| ldapUserFilterObjectclass     | person                                                                                                                                            |
| ldapUuidGroupAttribute        | auto                                                                                                                                              |
| ldapUuidUserAttribute         | auto                                                                                                                                              |
| turnOffCertCheck              | 0                                                                                                                                                 |
| useMemberOfToDetectMembership | 1                                                                                                                                                 |
+-------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------+
```

</details>

### Client configuration
**Browser:** Chrome 55

**Operating system:** Arch Linux

#### Web server error log
<details>
<summary>Web server error log</summary>
```
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>
```
{"reqId":"LO55aZHNLek71J7XWPqD","remoteAddr":"109.50.210.7","app":"user_ldap","message":"Bind failed: 49: Invalid credentials","level":2,"time":"2017-01-09T20:34:32+00:00","method":"PROPFIND","url":"\/remote.php\/dav\/calendars\/r3pek\/default\/","user":"--","version":"9.1.2.2"}
{"reqId":"LO55aZHNLek71J7XWPqD","remoteAddr":"109.50.210.7","app":"core","message":"Login failed: 'r3pek@r3pek.org' (Remote IP: '109.50.210.7')","level":2,"time":"2017-01-09T20:34:32+00:00","method":"PROPFIND","url":"\/remote.php\/dav\/calendars\/r3pek\/default\/","user":"--","version":"9.1.2.2"}
{"reqId":"BksS\/JFzdij\/cAa6otZP","remoteAddr":"109.50.210.7","app":"user_ldap","message":"Bind failed: 49: Invalid credentials","level":2,"time":"2017-01-09T20:34:33+00:00","method":"REPORT","url":"\/remote.php\/dav\/calendars\/r3pek\/default\/","user":"--","version":"9.1.2.2"}
{"reqId":"BksS\/JFzdij\/cAa6otZP","remoteAddr":"109.50.210.7","app":"core","message":"Login failed: 'r3pek@r3pek.org' (Remote IP: '109.50.210.7')","level":2,"time":"2017-01-09T20:34:33+00:00","method":"REPORT","url":"\/remote.php\/dav\/calendars\/r3pek\/default\/","user":"--","version":"9.1.2.2"}
{"reqId":"XewreyLKmNr\/N1IFgbnJ","remoteAddr":"109.50.210.7","app":"user_ldap","message":"Bind failed: 49: Invalid credentials","level":2,"time":"2017-01-09T20:34:33+00:00","method":"PROPFIND","url":"\/remote.php\/dav\/calendars\/r3pek\/work\/","user":"--","version":"9.1.2.2"}
{"reqId":"XewreyLKmNr\/N1IFgbnJ","remoteAddr":"109.50.210.7","app":"core","message":"Login failed: 'r3pek@r3pek.org' (Remote IP: '109.50.210.7')","level":2,"time":"2017-01-09T20:34:33+00:00","method":"PROPFIND","url":"\/remote.php\/dav\/calendars\/r3pek\/work\/","user":"--","version":"9.1.2.2"}
{"reqId":"cgk\/pQJYDiSxIK5T9\/3I","remoteAddr":"109.50.210.7","app":"user_ldap","message":"Bind failed: 49: Invalid credentials","level":2,"time":"2017-01-09T20:34:34+00:00","method":"REPORT","url":"\/remote.php\/dav\/calendars\/r3pek\/work\/","user":"--","version":"9.1.2.2"}
{"reqId":"cgk\/pQJYDiSxIK5T9\/3I","remoteAddr":"109.50.210.7","app":"core","message":"Login failed: 'r3pek@r3pek.org' (Remote IP: '109.50.210.7')","level":2,"time":"2017-01-09T20:34:34+00:00","method":"REPORT","url":"\/remote.php\/dav\/calendars\/r3pek\/work\/","user":"--","version":"9.1.2.2"}
{"reqId":"Io+F11u8RHqHGVRoomnc","remoteAddr":"109.50.210.7","app":"user_ldap","message":"Bind failed: 53: Server is unwilling to perform","level":2,"time":"2017-01-09T20:36:20+00:00","method":"GET","url":"\/ocs\/v2.php\/apps\/notifications\/api\/v1\/notifications?format=json","user":"r3pek","version":"9.1.2.2"}
{"reqId":"Io+F11u8RHqHGVRoomnc","remoteAddr":"109.50.210.7","app":"core","message":"Login failed: 'r3pek' (Remote IP: '109.50.210.7')","level":2,"time":"2017-01-09T20:36:20+00:00","method":"GET","url":"\/ocs\/v2.php\/apps\/notifications\/api\/v1\/notifications?format=json","user":"r3pek","version":"9.1.2.2"}
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>
```
```
</details>

**Comment by r3pek** [comment]
[nextcloud_debug.txt](https://github.com/nextcloud/server/files/694753/nextcloud_debug.txt)

Log file with Logging set to debugging. I started it right before a normal login with the main password and then forcing a sync on DAVDroid.

**Comment by ChristophWurst** [comment]
> Install and configure 2StepAuth plugin

Do you mean the TOTP app from https://apps.nextcloud.com/apps/twofactor_totp?

If an app password is used, we regularly (about once per 5 minutes) check the password. Maybe your LDAP server or the LDAP config blocks those password checks for some reason? Could that be the reason?

**Comment by r3pek** [comment]
Yes, that one. Hummmm, the App passwords are only defined on mysql and not on the LDAP backend. The problem is that check is being made against the LDAP backend *first* and only after to the App passwords DB, causing the other to be locked out (if such policy exist).

**Comment by ChristophWurst** [comment]
> App passwords are only defined on mysql and not on the LDAP backend

Yes, that's right. They are independent of the used user backend.

> The problem is that check is being made against the LDAP backend first and only after to the App passwords DB

I don't think so. If the app password is invalid (it does not exists), we do not query the user backend.

**Comment by r3pek** [comment]
Why do I have some many "Login failed" in the logs then?

**Comment by r3pek** [comment]
The LDAP server only locks on failed passwords, you can check 1000 times a seconds if you wanted to as long as they always valid ;)

My bad about the TOTP then, really thought app passwords as a thing from that plugin.

**Comment by ChristophWurst** [comment]
> The LDAP server only locks on failed passwords, you can check 1000 times a seconds if you wanted to as long as they always valid ;)

Good to know :)

@blizzz any idea?

**Comment by r3pek** [comment]
Just to make sure, just captured with tcpdump ;)
![screenshot from 2017-01-16 13-30-31](https://cloud.githubusercontent.com/assets/3050556/21984903/2fae14cc-dbf0-11e6-8fe8-6d282810bb18.png)

**Comment by blizzz** [comment]
@ChristophWurst 

> If the app password is invalid (it does not exists), we do not query the user backend.

And if the app password is valid, do you then query the backend?

**Comment by blizzz** [comment]
> Yes, we do query the user back-end about once per 5minutes to check if the login password has changed.

@ChristophWurst where in the source does this happen?

**Comment by blizzz** [comment]
on IRC @r3pek stated that the lock out happens within few minutes (using the desktop client). His LDAP Server locks the user after 6 failed login  attempts.

**Comment by blizzz** [comment]
> @ChristophWurst where in the source does this happen?

Nm, found it in Session::checkTokenCredentials()

**Comment by ChristophWurst** [comment]
Just saw the conversation on IRC. So what I missed before was that apparently the app password is sent to the user backend, which obviously fails. I do not yet understand how that is possible, because we actually try to validate the password as app password and perform the login password check only as fallback.

ref https://github.com/nextcloud/server/blob/182f570b046765fdb57d7eb513fe8c4b2490ed8f/lib/private/User/Session.php#L294-L300

**Comment by blizzz** [comment]
For the record, the Issue was that the used and stored login name did not match and multiple login names are not supported yet (https://github.com/nextcloud/server/blob/182f570b046765fdb57d7eb513fe8c4b2490ed8f/lib/private/User/Session.php#L659)

@ChristophWurst  Should we close it as "config" issue or can we do anything specifically about this? Is this check actually necessary?

**Comment by ChristophWurst** [comment]
> @ChristophWurst Should we close it as "config" issue or can we do anything specifically about this?

I will create new issue describing how we can remove this limitation. 

> Is this check actually necessary?

Unfortunately, yes, because the user backend's ``checkPassword`` expects the first parameter to be something like login name, e-mail or similar. We'd have to add a method that accepts uid + password.

**Comment by blizzz** [comment]
> Unfortunately, yes, because the user backend's checkPassword expects the first parameter to be something like login name, e-mail or similar. We'd have to add a method that accepts uid + password.

Well, the uid is a Nextcloud internal value… However, the backend could provide a method to resolve a loginname to a uid and hope that it works.  The LDAP Backend does it because of federated sharing needs, however we did not make it a general approach for the backend. This could be added (somehow, without breaking API? Throwing around with Interfaces?).

**Comment by ChristophWurst** [comment]
> Well, the uid is a Nextcloud internal value…

I was under the impression that that is our unique identifier for user objects across the whole system. We even have a getter for it in the public interface: https://github.com/nextcloud/server/blob/fa96b9e356cc52207a7a8ea5caf569203157ad1b/lib/public/IUser.php#L37-L43

**Comment by blizzz** [comment]
> I was under the impression that that is our unique identifier for user objects across the whole system.

Whole system? If you include other services, no, just Nextcloud. We also limit the available characters in the user id, so even  if a an external system provides an id it does not mean it is compatible with our constraints.

**Comment by ChristophWurst** [comment]
> Whole system? If you include other services, no, just Nextcloud. We also limit the available characters in the user id, so even if a an external system provides an id it does not mean it is compatible with our constraints.

Makes sense.

IIRC this issue was resolved  as by https://github.com/nextcloud/server/issues/2999#issuecomment-272960492. I'm closing this now.
